flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestDispatcherImpl.java

```diff
@@ -73,11 +79,27 @@ public void dispatch(ChannelStateWriteRequest request) throws Exception {
     private void dispatchInternal(ChannelStateWriteRequest request) throws Exception {
         if (request instanceof CheckpointStartRequest) {
             checkState(
-                    !writers.containsKey(request.getCheckpointId()),
-                    "writer not found for request " + request);
-            writers.put(request.getCheckpointId(), buildWriter((CheckpointStartRequest) request));
+                    request.getCheckpointId() > ongoingCheckpointId,
+                    String.format(
+                            "Checkpoint must be incremented, ongoingCheckpointId is %s, but the request is %s.",
+                            ongoingCheckpointId, request));
+            if (writer != null && !writer.isDone()) {
+                writer.fail(
+                        new IllegalStateException(
+                                String.format(
+                                        "Task[name=%s, subtaskIndex=%s] has uncompleted channelState writer of checkpointId=%s, "
+                                                + "but it received a new checkpoint start request of checkpointId=%s, it maybe "
+                                                + "a bug due to currently not supported concurrent unaligned checkpoint.",
+                                        taskName,
+                                        subtaskIndex,
+                                        ongoingCheckpointId,
+                                        request.getCheckpointId())));
+            }
```

Comment:
I think this might fail currently. I don't see `writer` being set to `null` if its checkpoint is aborted. Thus the following sequence I think will trigger this `IllegalStateException`.

1. start checkpoint 42
2. abort checkpoint 42
3. start checkpoint 43

Reply:
After step2 `abort checkpoint 42`, `writer.isDone()` will be true. So here won't throw exception during step3.

BTW, `writer` cannot be set to `null` during abort checkpoint. There may be some `writeInput/writeOutput` after abort. If set to `null`, `req.onWriterMissing();` will throw exception.

`writer` just be set to `null` when the checkpoint is completed, or `ChannelStateWriteRequestDispatcherImpl` failed.

Reply:
Hi @pnowojski , thanks for your review.

I have added the `ChannelStateWriterImplTest.testAbortOldAndStartNewCheckpoint()`. It is: 

1. start checkpoint 42
2. abort checkpoint 42
3. start checkpoint 43

And checked that result42 should fail due to `abort exception` instead of `new IllegalStateException()`,

Reply:
Hi @pnowojski , thanks for your suggestions.

I updated. I created the `CheckpointAbortRequest` and `maxAbortedCheckpointId`. I didn't use the `abortedCheckpointId`, because the abort request may be received in advance.

Reply:
If checkpoint 42 is completed, the `ongoingCheckpointId=42` and `writer == null`.

- `start(43)` has been added to the queue, but not be executed.
- `abort(44)` add the queue head and tail

When the dispatcher received the `abort(44)`, is the `abortedCheckpointId` 42?  Or the `isAborted` will be the true? 
 If yes, I prefer use `maxAbortedCheckpointId=44`. We can discard the `start(43)` directly. 


Also, `maxAbortedCheckpointId` is more friendly to channel state file merging. Multiple subtasks will share the same `ChannelStateWriteRequestDispatcherImpl` in the future, the order of requests is not guaranteed between multiple subtasks.  A subtask's checkpoint 44 is aborted, and other subtasks should also be aborted here. But other subtasks may also `start(42)`.

Do you think so?

Reply:
I was thinking about a contract that `ChannelStateWriter` is allowed to ignore any request that's not addressed to the `ongoingCheckpointId`. So in your scenario `abort(44)` could be safely ignored. Once `SubtaskCheckpointCoordinator` decides to abort/complete checkpoint `43` (for example it can be due to `notifyCheckpointAborted(43)` rpc, checkpoint `43` completed, or checkpoint `45` starting (and super-seeding `43`) it would take care of calling `ChannelStateWriter#abort(43)`.

> Also, maxAbortedCheckpointId is more friendly to channel state file merging.

Ok, that might be a good argument to keep it as it is.